# Lesson 15: Mathematical Operations, Equations and Sequences with NumPy Arrays

# HANDS-ON EXERCISE 1
## Business Analytics - Sales Performance Calculator

In [9]:
import numpy as np

# Task 1: Create the Dataset
daily_sales = np.array([125000, 150000, 175000, 140000, 190000, 210000, 160000])

print("Daily Sales Array:", daily_sales)
print("Type of the Array:", type(daily_sales))
print("Shape of the Array:", daily_sales.shape)

# Task 2: Calculate Total Sales
total_sales = np.sum(daily_sales)
print("Total Sales:", total_sales)

# Task 3: Calculate Average Daily Sales
average_sales = np.mean(daily_sales)
print("Average Daily Sales:", average_sales)

# Task 4: Apply a Sales Adjustment (10% increase)
adjusted_sales = daily_sales * 1.10
print("Adjusted Sales (10% increase):", adjusted_sales)

# Task 5: Compare Original and Adjusted Sales
sales_difference = adjusted_sales - daily_sales
print("Additional Revenue per Day:", sales_difference) 

# Task 6: Use NumPy's Mathematical Functions to find the average and then compare the result with the total sales 
average_sales = np.mean(daily_sales)
total_sales = np.sum(daily_sales)

# Write a short markdown explanation answering: What does the average daily sales figure tell the business that the total sales figure doess not

Daily Sales Array: [125000 150000 175000 140000 190000 210000 160000]
Type of the Array: <class 'numpy.ndarray'>
Shape of the Array: (7,)
Total Sales: 1150000
Average Daily Sales: 164285.7142857143
Adjusted Sales (10% increase): [137500. 165000. 192500. 154000. 209000. 231000. 176000.]
Additional Revenue per Day: [12500. 15000. 17500. 14000. 19000. 21000. 16000.]


#### Task 6: Write a short markdown explanation answering: What does the average daily sales figure tell the business that the total sales figure doess not
* **What the total sales tell us:** The total sales of 1,150,000 show the full amount of money the shop made during the whole week.

* **What the average daily sales tell us:** The average daily sales of 164,285.71 show the normal amount the shop makes on a typical day.

### Business Thinking Challenge
For this week, our total sales reached 1,150,000, giving us a clear picture of our overall revenue. Meanwhile, our average daily sales of 164,285.71 help us understand what a normal, typical day looks like for our shop. This helps us spot which days performed better than usual and plan better for the coming week.

# HANDS-ON EXERCISE 2
## Education Analytics - Student Performance Analysis

In [10]:
# Task 1: Create and Inspect the Data
scores = np.array([62, 75, 81, 69, 88, 94, 73, 85, 77, 91])
print("Scores:", scores)
print("Type:", type(scores))
print("Shape:", scores.shape)

# Task 2: Calculate the Mean Score
mean_score = np.mean(scores)
print("Mean Score:", mean_score)

# Task 3: Calculate Each Student's Deviation
deviations = scores - mean_score
print("Deviations:", deviations)

Scores: [62 75 81 69 88 94 73 85 77 91]
Type: <class 'numpy.ndarray'>
Shape: (10,)
Mean Score: 79.5
Deviations: [-17.5  -4.5   1.5 -10.5   8.5  14.5  -6.5   5.5  -2.5  11.5]


#### Task 4 Explanation: Identifying Deviations
* **What a positive deviation means:** It means the student scored above the class average.
* **What a negative deviation means:** It means the student scored below the class average.
* **What a value close to zero means:** It means the student scored very close to the class average.

In [11]:
# Task 5: Square the Deviations
squared_deviations = deviations ** 2
print("Squared Deviations:", squared_deviations)

Squared Deviations: [306.25  20.25   2.25 110.25  72.25 210.25  42.25  30.25   6.25 132.25]


#### Task 5 Explanation: Why Square the Deviations?
Squaring turns all negative values into positive values because multiplying any negative number by itself always gives a positive result. This prevents negative and positive deviations from canceling each other out when we add them up.

In [12]:
# Task 6: Calculate the Sum of Squared Deviations
sum_squared_deviations = np.sum(squared_deviations)
print("Sum of Squared Deviations:", sum_squared_deviations)

Sum of Squared Deviations: 932.5


#### Task 7: Standard Deviation Process and Usefulness

**The Conceptual Process:**
1. Calculate the mean
2. Calculate deviations
3. Square the deviations
4. Sum the squared deviations
5. Continue with the appropriate division
6. Take the square root

**Why Standard Deviation is Useful:** Standard deviation is useful because it tells you how spread out a group of numbers is from the average.

* **Low Standard Deviation:** Most students scored close to the class average, showing that the class performed very evenly.
* **High Standard Deviation:** The scores are scattered, meaning some students did extremely well while others struggled.

It gives the program coordinator a clear picture of the overall consistency of the class rather than just looking at the average alone.


### Industry Thinking Challenge

While the average score gives us a general overview, it can hide important details about the class. Looking at how individual scores deviate from the average helps us see if the students are performing consistently or if some students are falling behind and need extra help while others are excelling.



# HANDS-ON EXERCISE 3
## Engineering & Scientific Computing - Trigonometric Series and Convergence

In [15]:
import numpy as np

# Task 1: Define the Angle and Convert to Radians
theta = 30
theta_radians = np.radians(theta)
print("Theta in Radians:", theta_radians)

# Task 2: Create the Sequence of Terms (k = 1 to 99)
k = np.arange(1, 100, 1)
print("Array k:", k)
print("First Value:", k[0])
print("Last Value:", k[-1])
print("Number of Values Generated:", len(k))

# Task 3: Calculate the Trigonometric Component
sine_component = np.sin(theta_radians)
print("Sine Component:", sine_component)

# Task 4: Build the Series Terms
series_terms = sine_component / k
print("Series Terms:", series_terms)

# Task 5: Calculate the Sum
series_sum = np.sum(series_terms)
print("Series Sum (for 99 terms):", series_sum)

Theta in Radians: 0.5235987755982988
Array k: [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72
 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96
 97 98 99]
First Value: 1
Last Value: 99
Number of Values Generated: 99
Sine Component: 0.49999999999999994
Series Terms: [0.5        0.25       0.16666667 0.125      0.1        0.08333333
 0.07142857 0.0625     0.05555556 0.05       0.04545455 0.04166667
 0.03846154 0.03571429 0.03333333 0.03125    0.02941176 0.02777778
 0.02631579 0.025      0.02380952 0.02272727 0.02173913 0.02083333
 0.02       0.01923077 0.01851852 0.01785714 0.01724138 0.01666667
 0.01612903 0.015625   0.01515152 0.01470588 0.01428571 0.01388889
 0.01351351 0.01315789 0.01282051 0.0125     0.01219512 0.01190476
 0.01162791 0.01136364 0.01111111 0.01086957 0.0106383  0.01041667
 0.0

In [16]:
# Task 6: Experimenting with larger numbers of terms

# 1. For 99 terms (already done above, but let's store it cleanly)
k_99 = np.arange(1, 100, 1)
sum_99 = np.sum(np.sin(theta_radians) / k_99)

# 2. For 999 terms
k_999 = np.arange(1, 1000, 1)
sum_999 = np.sum(np.sin(theta_radians) / k_999)

# 3. For 9,999 terms
k_9999 = np.arange(1, 10000, 1)
sum_9999 = np.sum(np.sin(theta_radians) / k_9999)

print("Sum with 99 terms:", sum_99)
print("Sum with 999 terms:", sum_999)
print("Sum with 9,999 terms:", sum_9999)

Sum with 99 terms: 2.5886887588198104
Sum with 999 terms: 3.7422354302751724
Sum with 9,999 terms: 4.89375301802219


#### Task 7: Investigate Convergence Table

| Number of Terms | Series Result |
| :--- | :--- |
| 99 | 2.588687588198104 |
| 999 | 3.7422354302751724 |
| 9,999 | 4.89375301802219 |

**What happens to the result as the number of terms increases?**
As the number of terms increases, the total sum continues to grow larger.

#### Task 8: Explanation of Convergence

1. **What happened when you used fewer terms?**
When we used a small number of terms like 99, the total sum was small, giving us an answer of approximately 2.6. Because we only added a few numbers together, the calculation stopped early and only captured a tiny piece of the overall sequence.

2. **What happened when you increased the number of terms?**
When we increased the terms to 999, the total sum grew larger to approximately 3.74. And when we pushed it further to 9,999 terms, the total sum increased even more to reach approximately 4.9. As we kept feeding more numbers into our code, the running total continued to rise because every new term added its own value to the final pile.

3. **Why does increasing the number of terms change the approximation?**
It changes the answer because every new term we include brings another fraction into our math problem. When we use a small range like 99, our answer is just a basic estimate. But when we include thousands of terms up to 9,999, our code looks at a much bigger picture, which gives us a much more complete and thorough final calculation.

4. **What does it mean when the result gets closer to a limiting value?**
It means we are studying how our math series behaves over the long run. Even though our sums grew from 2.6 to 3.7 and 4.9 in this experiment, watching how these numbers change helps us understand how stable, reliable, and predictable our computer code is when handling large amounts of data.







# Bonus Challenge: Farm Crop Yield Monitoring Experiment

* **Scenario:** Agriculture
* **Objective:** To use NumPy arrays and basic calculations to analyze, track and manage weekly farm harvest performance.

### Dataset
We have a NumPy array representing the weekly crop harvest yields (in kilograms) recorded across four weeks:
* Week 1: 120 kg
* Week 2: 135 kg
* Week 3: 150 kg
* Week 4: 165 kg

In [18]:
import numpy as np

# 1. Our Numerical Dataset
harvest_weights = np.array([120, 135, 150, 165])

# --- SIX CALCULATIONS ---

# Calculation 1: Total seasonal yield
total_yield = np.sum(harvest_weights)

# Calculation 2: Average weekly yield
average_yield = np.mean(harvest_weights)

# Calculation 3: Scaling growth up by 10%
boosted_yield = np.multiply(harvest_weights, 1.1)

# Calculation 4: Accounting for a 5 kg post-harvest loss
net_harvest = np.subtract(harvest_weights, 5.0)

# Calculation 5: Adding a 15 kg cooperative bonus
subsidy_added = np.add(harvest_weights, 15.0)

# Calculation 6: Splitting the harvest equally between two markets
market_split = np.divide(harvest_weights, 2.0)

# --- PRINT RESULTS CLEARLY ---
print("--- Farm Harvest Analysis ---")
print("Original Harvests:", harvest_weights)
print("Calculation 1 - Total Yield:", total_yield, "kg")
print("Calculation 2 - Average Yield:", average_yield, "kg")
print("Calculation 3 - Boosted Yield (10% Up):", boosted_yield)
print("Calculation 4 - Net Harvest (After 5kg Loss):", net_harvest)
print("Calculation 5 - Subsidy Added (+15kg Bonus):", subsidy_added)
print("Calculation 6 - Market Split (Two Locations):", market_split)

--- Farm Harvest Analysis ---
Original Harvests: [120 135 150 165]
Calculation 1 - Total Yield: 570 kg
Calculation 2 - Average Yield: 142.5 kg
Calculation 3 - Boosted Yield (10% Up): [132.  148.5 165.  181.5]
Calculation 4 - Net Harvest (After 5kg Loss): [115. 130. 145. 160.]
Calculation 5 - Subsidy Added (+15kg Bonus): [135. 150. 165. 180.]
Calculation 6 - Market Split (Two Locations): [60.  67.5 75.  82.5]


### Interpretation of Results

* **Calculation 1 (Total Seasonal Yield):**
By using `np.sum()`, we added all the weekly harvest weights together (120, 135, 150 and 165). The total production reached 570 kg across the four weeks, giving the farm manager a clear picture of the overall harvest.

* **Calculation 2 (Average Weekly Yield):**
Using `np.mean()`, we found the average weight across the four weeks. The average output came to 142.5 kg per week, giving us a good baseline for normal weekly production.

* **Calculation 3 (Growth Projection):**
Using `np.multiply()`, we increased each week's harvest by 10% (multiplying by 1.1). This helps us project what the output will look like if weather and farming conditions improve next season.

* **Calculation 4 (Waste Adjustment):**
Using `np.subtract()`, we took away 5 kg from each week. This accounts for normal post-harvest losses, like damaged crops or waste during transport.

* **Calculation 5 (Bonus Subsidy):**
Using `np.add()`, we added 15 kg to each week's record. This simulates an extra supply boost or fertilizer bonus given to the farmers.

* **Calculation 6 (Market Distribution):**
Using `np.divide()`, we split each week's weight into two equal parts. This helps logistics planners see how to share the food stock evenly between two different sales markets.

### Conclusion
This project successfully used NumPy tools (`np.array`, `np.sum`, `np.mean`, `np.multiply`, `np.subtract`, `np.add`, and `np.divide`) to solve real-world farming challenges. It shows how code can make tracking, analyzing, and planning farm production fast and easy.


